egs/csj/s5/local/nnet/run_cnn.sh

```diff
@@ -0,0 +1,136 @@
+    $gmm/graph_csj_tg data-fbank/$eval_num $dir/decode_${eval_num}_csj || exit 1;
+  done
+fi
+
+# Pre-train stack of RBMs on top of the convolutional layers (4 layers, 1024 units),
+if [ $stage -le 2 ]; then
+  dir=exp/cnn4_pretrain-dbn
+  transf_cnn=exp/cnn4/final.feature_transform_cnn # transform with convolutional layers
+  # Train
+  $cuda_cmd $dir/log/pretrain_dbn.log \
+    steps/nnet/pretrain_dbn.sh --nn-depth 4 --hid-dim 1024 --rbm-iter 20 \
+    --feature-transform $transf_cnn --input-vis-type bern \
+    --param-stddev-first 0.05 --param-stddev 0.05 \
+    $train $dir || exit 1
+fi
+
+# Re-align using CNN,
+if [ $stage -le 3 ]; then
+  dir=exp/cnn4
+  steps/nnet/align.sh --nj 10 --cmd "$train_cmd" \
+    $train data/lang $dir ${dir}_ali || exit 1
+fi
+
+# Train the DNN optimizing cross-entropy,
+if [ $stage -le 4 ]; then
+  dir=exp/cnn4_pretrain-dbn_dnn; [ ! -d $dir ] && mkdir -p $dir/log;
+  ali=exp/cnn4_ali
+  feature_transform=exp/cnn4/final.feature_transform
+  feature_transform_dbn=exp/cnn4_pretrain-dbn/final.feature_transform
+  dbn=exp/cnn4_pretrain-dbn/4.dbn
+  cnn_dbn=$dir/cnn_dbn.nnet
+  { # Concatenate CNN layers and DBN,
+    num_components=$(nnet-info $feature_transform | grep -m1 num-components | awk '{print $2;}')
+    cnn="nnet-copy --remove-first-layers=$num_components $feature_transform_dbn - |"
+    nnet-concat "$cnn" $dbn $cnn_dbn 2>$dir/log/concat_cnn_dbn.log || exit 1 
+  }
+  # Train
+  $cuda_cmd $dir/log/train_nnet.log \
+    steps/nnet/train.sh --feature-transform $feature_transform --dbn $cnn_dbn --hid-layers 0 \
+    ${train}_tr90 ${train}_cv10 data/lang $ali $ali $dir || exit 1;
+  # Decode (reuse HCLG graph)
+  for eval_num in eval1 eval2 eval3 $dev_set ;do
+      steps/nnet/decode.sh --nj 10 --cmd "$decode_cmd" --config conf/decode_dnn.config --acwt 0.2 \
+    $gmm/graph_csj_tg data-fbank/$eval_num $dir/decode_${eval_num}_csj || exit 1;
+  done
+fi
+
+# Sequence training using sMBR criterion, we do Stochastic-GD 
+# with per-utterance updates. For RM good acwt is 0.2,
+dir=exp/cnn4_pretrain-dbn_dnn_smbr
+srcdir=exp/cnn4_pretrain-dbn_dnn
+acwt=0.2
+
+# First we generate lattices and alignments,
+if [ $stage -le 4 ]; then
+  steps/nnet/align.sh --nj 10 --cmd "$train_cmd" \
+    $train data/lang $srcdir ${srcdir}_ali || exit 1;
+  steps/nnet/make_denlats.sh --nj 10 --cmd "$decode_cmd" --config conf/decode_dnn.config --acwt $acwt \
+    $train data/lang $srcdir ${srcdir}_denlats || exit 1;
+fi
+
+# Re-train the DNN by 6 iterations of sMBR,
+if [ $stage -le 5 ]; then
+  steps/nnet/train_mpe.sh --cmd "$cuda_cmd" --num-iters 6 --acwt $acwt --do-smbr true \
+    $train data/lang $srcdir ${srcdir}_ali ${srcdir}_denlats $dir || exit 1
+  # Decode
+  for ITER in 1 2 3 4 5 6; do
+    steps/nnet/decode.sh --nj 10 --cmd "$decode_cmd" --config conf/decode_dnn.config \
+      --nnet $dir/${ITER}.nnet --acwt $acwt \
+      $gmm/graph data-fbank/$eval_num $dir/decode_${eval_num}_csj_it${ITER} || exit 1
+  done 
+fi
```

Comment:
@sw005320 Shinji, how big is the training set? Seems like the 'nnet1' scripts come from egs/rm/s5/local/nnet. It might be fine to use only 4 epochs of sMBR instead of 6.